### PR TITLE
feat(costModels): add the create-cost-model wizard

### LINKED ISSUE #882: Price list: create
## User Story
As user, I want to create a price list so I can see what my OCP environment costs.

## Impacts
UI, Docs

## Role
User

## Assumptions
- Koku API has support to create price lists

## UI Details
See mocks:
https://redhat.invisionapp.com/share/NJS0RS3Q7KE#/screens/364417349

Screen Shot 2019-05-28 at 12 04 18 PM

## Acceptance Criteria
- [ ] A user should be able to create a new Rate, modify an existing Rate, and delete Rates.
- [ ] A `Rate` should define a `Cost` that applies to a `Metric` and a `Time`. (e.g. $0.25 per CPU-hour, or $0.01 per GiB Network-IO per minute.)
- [ ] A user should be able to associate a Rate with an Account or OCP Cluster

### PR DESCRIPTION
**Pull request by boaz0**
closes #882

### CODE REVIEW

**Comment by dlabrecq** [comment]
I built the source using `BRANCH=prod-beta`, but see a black screen when navigating to https://ci.foo.redhat.com:1337/hybrid/cost-management/cost-models.

However, I retested #926 and that build still works.

I'll review again when #926 is merged, so it's easier to see the latest code changes here.

**Comment by boaz0** [comment]
Removed the FilterIcon as suggested.

**Comment by dlabrecq** [comment]
Please let me know when you rebase with the latest changes and I'll review again?

File: src/pages/createCostModelWizard/markup.tsx [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+              <TextInput
+                isRequired
+                type="text"
+                id="markup"
+                name="markup"
+                value={markup}
+                onChange={onMarkupChange}
+                isValid={isValidMarkup}
+              />
+              <InputGroupText>%</InputGroupText>
```
Comment (dlabrecq) [inline src/pages/createCostModelWizard/markup.tsx]: Similar issue as the filter icon. That is, the percentage appears outside the text field, surrounded by a border. The mockup shows something more like this:

Screen Shot 2019-08-16 at 10 39 52 AM

File: src/pages/createCostModelWizard/filterToolbar.tsx [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+                      type,
+                      addMultiValueQuery(query)('name', filterName),
+                      page,
+                      perPage
+                    );
+                  }}
+                />
+                <InputGroupText>
+                  <SearchIcon />
+                </InputGroupText>
```
Comment (dlabrecq) [inline src/pages/createCostModelWizard/filterToolbar.tsx]: Similar issue as the filter icon. That is, the search icon appears outside the text field, surrounded by a border. The mockup shows something more like this:

Screen Shot 2019-08-16 at 10 45 56 AM

File: src/pages/createCostModelWizard/filterToolbar.tsx [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        type,
+        fetchSources,
+      }) => {
+        return (
+          <ToolbarGroup>
+            <ToolbarItem>
+              <InputGroup>
+                <InputGroupText>
+                  <FilterIcon />
+                </InputGroupText>
```
Comment (dlabrecq) [inline src/pages/createCostModelWizard/filterToolbar.tsx]: Similar issue as the filter icon shown in the cost model page. That is, the search icon appears outside the text field, surrounded by a border. The mockup shows something more like this:

Screen Shot 2019-08-16 at 10 45 56 AM

File: src/pages/createCostModelWizard/addPriceList.tsx [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+              <FormGroup label="Rate" fieldId="rate-input-box">
+                <InputGroup>
+                  <TextInput
+                    type="number"
+                    aria-label="rate in dollars"
+                    id="rate-input-box"
+                    value={priceListCurrent.rate}
+                    onChange={(value: string) => updateCurrentPL('rate', value)}
+                  />
+                  <InputGroupText>$</InputGroupText>
```
Comment (dlabrecq) [inline src/pages/createCostModelWizard/addPriceList.tsx]: The $ symbol appears outside the text field, surrounded by a border. The mockup shows something more like this:

Screen Shot 2019-08-16 at 10 39 52 AM

File: src/pages/createCostModelWizard/priceListTable.tsx [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+                            }
+                            this.setState({
+                              filter: this.state.current,
+                              current: '',
+                            });
+                          }}
+                        />
+                        <InputGroupText>
+                          <SearchIcon />
+                        </InputGroupText>
```
Comment (dlabrecq) [inline src/pages/createCostModelWizard/priceListTable.tsx]: Similar issue as the filter icon. That is, the search icon appears outside the text field, surrounded by a border. The mockup shows something more like this:

Screen Shot 2019-08-16 at 10 45 56 AM

Comment (dlabrecq) [reply src/pages/createCostModelWizard/priceListTable.tsx]: Also want to note that the filter in the cost model page does not have the search icon.

**Comment by boaz0** [comment]
Hi @dlabrecq I fixed the problems you mentioned. Please let me know if I missed something or I didn't fix it correctly.

Thanks.

**Comment by dlabrecq** [comment]
Discovered a couple missing localizations.

src/pages/createCostModelWizard/generalInformation.tsx

Screen Shot 2019-08-20 at 4 39 17 PM

src/pages/createCostModelWizard/index.tsx

Screen Shot 2019-08-20 at 4 40 39 PM

**Comment by dlabrecq** [comment]
FYI, I created an issue for the required asterisk. Please correct me if I'm wrong, but that looks like a PatternFly issue?

https://github.com/patternfly/patternfly-react/issues/2746

**Comment by boaz0** [comment]
@dlabrecq -

- I will fix the missing localizations.

- I thought the "optional" text in the title was good enough because it's not mandatory (unless you select a metric and then the "Next" button is disabled until you add a rate). If you don't set price list it is an empty list.

Regarding the red asterisk I fixed it and I can see it on Firefox. What browser are you using?

image

Thanks.

**Comment by dlabrecq** [comment]
The "optional" text is very helpful. I simply meant that a default rate of 0 would be helpful as well. It would help clarify what type of value the user needs to enter in that input.

Screen Shot 2019-08-21 at 9 20 18 AM

**Comment by boaz0** [comment]
@dlabrecq hoo I see. No problem.

**Comment by boaz0** [comment]
@dlabrecq updated PR to have a default value in rate.

**Comment by boaz0** [comment]
Thanks a lot @dlabrecq !

**Pull request merged**